File: client/api/ws_client.py
"""
Unified WebSocket client.

A single persistent connection handles both room messages and user-level
notifications.  The server already supports this: room_id is optional on
connect, and the server subscribes the socket to a room the first time it
receives a message for that room (or via the initial query param).

Usage
-----
    client = UnifiedWsClient(
        token="...",
        on_room_message=handle_msg,       # called for type=="message" / "encrypted_message"
        on_notification=handle_notif,     # called for type=="invite" / "member_joined" / etc.
        on_reconnecting=handle_reconnect,
    )
    # Optionally subscribe to a room on connect:
    client.set_room(room_id)
    await client.connect()          # runs forever (call in a task)

    # Switch rooms without reconnecting:
    client.set_room(new_room_id)

    # Send a chat message:
    await client.send_message(room_id, body, files)

    # Shut down:
    client.close()

Backwards-compatible shims
--------------------------
WsClient and NotificationClient are kept as thin wrappers so existing
call-sites in room_view / chat_list_view / room_list_view continue to work
without changes.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Callable

import websockets
import websockets.exceptions

from config import WS_URL

logger = logging.getLogger(__name__)


class UnifiedWsClient:
    """Single WebSocket connection for both room messages and notifications."""

    _INITIAL_DELAY = 1.0
    _MAX_DELAY = 30.0

    def __init__(
        self,
        token: str,
        on_room_message: Callable[[dict], None] | None = None,
        on_notification: Callable[[dict], None] | None = None,
        on_reconnecting: Callable[[float], None] | None = None,
        ws_url: str | None = None,
    ) -> None:
        self._token = token
        self._on_room_message = on_room_message
        self._on_notification = on_notification
        self._on_reconnecting = on_reconnecting
        self._ws_url = ws_url or WS_URL
        self._closed = False
        self._ws = None
        # Current room the client is subscribed to (sent as query param on connect)
        self._room_id: int | None = None

    def set_room(self, room_id: int | None) -> None:
        """Set the room to subscribe to on the next (re)connect."""
        self._room_id = room_id

    async def connect(self) -> None:
        """Connect and loop forever with exponential back-off on failure."""
        delay = self._INITIAL_DELAY
        while not self._closed:
            url = f"{self._ws_url}?token={self._token}"
            if self._room_id is not None:
                url += f"&room_id={self._room_id}"
            try:
                async with websockets.connect(url) as ws:
                    self._ws = ws
                    delay = self._INITIAL_DELAY
                    logger.debug("[WS] Connected (room_id=%s)", self._room_id)
                    async for raw in ws:
                        if self._closed:
                            break
                        self._dispatch(raw)
            except (
                websockets.exceptions.ConnectionClosed,
                websockets.exceptions.WebSocketException,
                OSError,
            ) as exc:
                logger.debug("[WS] Connection lost: %s", exc)
            finally:
                self._ws = None

            if self._closed:
                break

            if self._on_reconnecting is not None:
                self._on_reconnecting(delay)

            logger.debug("[WS] Reconnecting in %.1fs", delay)
            await asyncio.sleep(delay)
            delay = min(delay * 2, self._MAX_DELAY)

    def _dispatch(self, raw: str) -> None:
        try:
            payload = json.loads(raw)
        except Exception:
            return

        msg_type = payload.get("type", "")

        # Room-level frames
        if msg_type in ("message", "encrypted_message"):
            if self._on_room_message is not None:
                try:
                    self._on_room_message(payload)
                except Exception as exc:
                    logger.error("[WS] on_room_message raised: %s", exc, exc_info=True)
        else:
            # Everything else is a notification (invite, member_joined, error, …)
            if self._on_notification is not None:
                try:
                    self._on_notification(payload)
                except Exception as exc:
                    logger.error("[WS] on_notification raised: %s", exc, exc_info=True)

    async def send_message(
        self,
        room_id: int,
        body: str,
        files: list[dict] | None = None,
    ) -> None:
        if self._ws is None:
            logger.warning("[WS] send_message called but not connected")
            return
        frame: dict = {"type": "message", "room_id": room_id, "body": body}
        if files:
            frame["files"] = files
        await self._ws.send(json.dumps(frame))

    def close(self) -> None:
        """Signal the client to stop and close the underlying socket."""
        self._closed = True
        if self._ws is not None:
            try:
                loop = asyncio.get_event_loop()
                if loop.is_running():
                    loop.create_task(self._ws.close())
            except RuntimeError:
                pass


# ---------------------------------------------------------------------------
# Backwards-compatible shims
# ---------------------------------------------------------------------------

class WsClient(UnifiedWsClient):
    """Shim: behaves like the old WsClient (room-only messages)."""

    def __init__(
        self,
        token: str,
        room_id: int,
        on_message: Callable[[dict], None],
        on_reconnecting: Callable[[float], None] | None = None,
        ws_url: str | None = None,
    ) -> None:
        super().__init__(
            token=token,
            on_room_message=on_message,
            on_reconnecting=on_reconnecting,
            ws_url=ws_url,
        )
        self.set_room(room_id)


class NotificationClient(UnifiedWsClient):
    """Shim: behaves like the old NotificationClient (notifications only)."""

    def __init__(
        self,
        token: str,
        on_notification: Callable[[dict], None],
        ws_url: str | None = None,
    ) -> None:
        super().__init__(
            token=token,
            on_notification=on_notification,
            ws_url=ws_url,
        )

File: flutter_client/lib/api/ws_client.dart
/// Unified WebSocket client (mirrors the Python client's `ws_client.py`).
///
/// A single persistent connection handles both room messages and user-level
/// notifications. The server subscribes the socket to a room from the room_id
/// query param at connect time.
library;

import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

typedef WsRoomMessageHandler = void Function(Map<String, dynamic> payload);
typedef WsNotificationHandler = void Function(Map<String, dynamic> payload);
typedef WsReconnectingHandler = void Function(double delay);

class UnifiedWsClient {
  UnifiedWsClient({
    required this.token,
    this.onRoomMessage,
    this.onNotification,
    this.onReconnecting,
    this.wsUrl,
  });

  final String token;
  WsRoomMessageHandler? onRoomMessage;
  WsNotificationHandler? onNotification;
  WsReconnectingHandler? onReconnecting;
  final String? wsUrl;

  static const double _initialDelay = 1.0;
  static const double _maxDelay = 30.0;

  WebSocketChannel? _channel;
  int? _roomId;
  bool _closed = false;

  int? get roomId => _roomId;

  void setRoom(int? roomId) => _roomId = roomId;

  Future<void> connect(String baseWsUrl) async {
    var delay = _initialDelay;
    while (!_closed) {
      var url = '$baseWsUrl?token=$token';
      if (_roomId != null) url += '&room_id=$_roomId';
      try {
        final channel = WebSocketChannel.connect(Uri.parse(url));
        _channel = channel;
        delay = _initialDelay;
        await for (final raw in channel.stream) {
          if (_closed) break;
          _dispatch(raw.toString());
        }
      } catch (_) {
        // connection lost; fall through to backoff
      } finally {
        _channel = null;
      }
      if (_closed) break;
      onReconnecting?.call(delay);
      await Future.delayed(Duration(milliseconds: (delay * 1000).toInt()));
      delay = (delay * 2).clamp(_initialDelay, _maxDelay);
    }
  }

  void _dispatch(String raw) {
    Map<String, dynamic>? payload;
    try {
      payload = jsonDecode(raw) as Map<String, dynamic>;
    } catch (_) {
      return;
    }
    final type = payload['type'] as String? ?? '';
    if (type == 'message' || type == 'encrypted_message') {
      onRoomMessage?.call(payload);
    } else {
      onNotification?.call(payload);
    }
  }

  void sendMessage(int roomId, String body, {List<Map<String, dynamic>>? files}) {
    if (_channel == null) return;
    final frame = {
      'type': 'message',
      'room_id': roomId,
      'body': body,
      if (files != null) 'files': files,
    };
    _channel!.sink.add(jsonEncode(frame));
  }

  void close() {
    _closed = true;
    _channel?.sink.close();
  }
}
